### ISSUE TITLE
consider adding `strict_top_level_inference`

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/core#836]
We should consider adding a `strict_top_level_inference` lint, as described here: https://github.com/dart-lang/sdk/issues/59562. Note that this lint does not yet exist.

This lint would warn about every omitted return type, parameter type, and variable type of a top-level declaration or class-like-namespace-level declaration (static or instance member- or constructor declaration), which is not given a type by inference, and which therefore defaults to dynamic.

**Comment by devoncarew** [comment]
Some details:

- originally landed in Dec 2024; it shipped in 3.7 stable
- it’s been used in the `dart_flutter_team_lints` set; most feedback was around wildcards
- it does have a quick fix?

cc @srawlins to keep me honest for the above

@natebosch, @lrhn, @chunhtai - can you review this lint? This was one we'd requested, so I'd guess mostly we'd be considering if there are any open issues to address, timing of landing it, which lint set, ... But feel free to do a general review.

**Comment by lrhn** [comment]
I like it. It's a way to avoid accidental `dynamic` in top-level declarations. 

The wildcard parameters should probably be exempt. Then :+1:.

**Comment by natebosch** [comment]
I am in favor of including this lint in the recommended set as it is currently implemented.

> The wildcard parameters should probably be exempt. Then 👍.

I initially leaned towards exempting wildcard parameters too, but I was convinced by you that it was better not to make them exempt 😄 

https://github.com/dart-lang/sdk/issues/59562#issuecomment-2532134637

> > Can a wildcard parameter have an omitted type?
> 
> I'd say no. This is not about having a type inside the function, so whether the parameter can be referenced or not doesn't matter. It's about making sure that the parameter has a type in the API, and doesn't default to `dynamic`.

**Comment by natebosch** [comment]
Oh, but we also changed course on that and wildcards _are_ exempt now?

https://github.com/dart-lang/sdk/commit/d5981addc0a7d59ffe7b55e75dd29c1c5bb494fa

That's also fine and I'm happy to still include the lint as it is currently implemented 😄

**Comment by srawlins** [comment]
Yes, the lint rule does have a quick fix. The lint rule shipped in Dart 3.7.

**Comment by devoncarew** [comment]
After discussion we do want to add this to our lint sets, and into the `core` set (so, this _will_ affect a package's pub score).

We also do want a bit more user feedback on this one (anything unexpected in the implementation?), so will likely ask the community to try it out.

**Comment by parlough** [comment]
I'm all for this inclusion! However:

> We also do want a bit more user feedback on this one (anything unexpected in the implementation?), so will likely ask the community to try it out.

Perhaps it should be in the recommended set for a cycle alongside an announcement of its eventual graduation to core?

**Comment by devoncarew** [comment]
> > We also do want a bit more user feedback on this one (anything unexpected in the implementation?), so will likely ask the community to try it out.
> 
> Perhaps it should be in the recommended set for a cycle alongside an announcement of its eventual graduation to core?

This is a good idea. However, we likely wouldn't ship the next stable version of this rule set until the April / May timeframe (we try not to ship new stable version of this package more frequently than every 6 months), so we have a bit of time to get some feedback.

**Comment by parlough** [comment]
> However, we likely wouldn't ship the next stable version of this rule set until the April / May timeframe.

I see. Then I think a `-dev` release with it in `core` for developers to prepare for it with some small announcements to increase awareness would be sufficient :)

I could call out the dev release in the various package maintainers Discord channels and perhaps pub.dev could surface a note on the scoring page when the latest release of the `lints` package is a `-dev` release?
